File: samples/bot-conversation-quickstart/csharp/BotConversationQuickStart/Bots/BotActivityHandler.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using Microsoft.Bot.Builder;
using Microsoft.Bot.Schema;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder.Teams;
namespace BotConversationQuickStart.Bots
{
    public class BotActivityHandler : TeamsActivityHandler
    {
        public BotActivityHandler() { }


        /// <summary>
        /// Handles incoming message activities from the user.
        /// If the message text is "Hello", it triggers a response via MentionActivityAsync.
        /// Otherwise, it sends a Hero Card with a "Say Hello" button using MessageBack action.
        /// </summary>
        /// <param name="turnContext">Context object containing information about the incoming message activity.</param>
        /// <param name="cancellationToken">Token for cancelling the async operation.</param>
        /// <returns>A task that represents the asynchronous operation.</returns>
        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
        {
            var userMessage = turnContext.Activity.Text?.Trim();

            if (userMessage == "Hello")
            {
                await MentionActivityAsync(turnContext, cancellationToken);
            }
            else
            {
                var card = new HeroCard
                {
                    Title = "Let's talk...",
                    Buttons = new List<CardAction>
                    {
                        new CardAction
                        {
                            Type = ActionTypes.MessageBack,
                            Title = "Say Hello",
                            Text = "Hello",
                            Value = new { count = 0 }
                        }
                    }
                };

                // Sends the Hero Card with the "Say Hello" button as a response to the user.
                await turnContext.SendActivityAsync(MessageFactory.Attachment(card.ToAttachment()), cancellationToken);
            }
        }

        /// <summary>
        /// Sends a mention activity that tags the user in the reply message.
        /// Constructs a mention entity using the user's name and includes it in the response.
        /// </summary>
        /// <param name="turnContext">Context object containing information about the incoming message activity.</param>
        /// <param name="cancellationToken">Token to cancel the asynchronous operation if needed.</param>
        /// <returns>A task that represents the asynchronous operation.</returns>
        private async Task MentionActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
        {
            var user = turnContext.Activity.From;

            var mention = new Mention
            {
                Mentioned = user,
                Text = $"<at>{user.Name}</at>",
                Type = "mention"
            };

            var reply = MessageFactory.Text($"Hi {mention.Text}");
            reply.Entities = new List<Entity> { mention };

            await turnContext.SendActivityAsync(reply, cancellationToken);
        }

    }
}

File: samples/bot-conversation-quickstart/python/bots/botActivityHandler.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from botbuilder.core import TurnContext, MessageFactory
from botbuilder.schema import HeroCard, CardAction, ActionTypes, Mention, Attachment
from botbuilder.core.teams import TeamsActivityHandler
from html import escape

class BotActivityHandler(TeamsActivityHandler):
    def __init__(self):
        super().__init__()

        # Registers an activity event handler for the message activity event.
        # This is triggered whenever the bot receives a message.
        self.on_message_activity = self.handle_message

    async def handle_message(self, turn_context: TurnContext):
        """
        Handles incoming messages sent to the bot.

        Args:
            turn_context (TurnContext): Provides context for the current turn of conversation.
        """
        # Removes any mention of the bot in the received message text (e.g., "@BotName Hello").
        turn_context.remove_recipient_mention(turn_context.activity)

        # Extract the user's message text and strip any extra spaces.
        user_message = turn_context.activity.text.strip()

        # If the user's message is "Hello", respond with a personalized mention.
        if user_message == "Hello":
            await self.mention_activity_async(turn_context)
        else:
            # If the user's message is something else, respond with a hero card.
            value = {"count": 0}  # Example value to be passed with the card action.
            card = HeroCard(
                title="Let's talk...",  # Title of the card.
                buttons=[
                    # Add a button to the card that sends a "Hello" message back to the bot.
                    CardAction(
                        type=ActionTypes.message_back,  # Button type: sends a message back.
                        title="Say Hello",  # Button text displayed to the user.
                        value=value,  # Payload sent when the button is clicked.
                        text="Hello",  # Message text sent when the button is clicked.
                    )
                ],
            )
            # Convert the hero card to an attachment that can be sent in the bot's response.
            attachment = Attachment(
                content_type="application/vnd.microsoft.card.hero",  # Indicates this is a hero card.
                content=card  # The actual hero card content.
            )
            # Send the hero card as the bot's response.
            await turn_context.send_activity(
                MessageFactory.attachment(attachment)
            )

    async def mention_activity_async(self, turn_context: TurnContext):
        """
        Sends a reply that mentions the user.

        Args:
            turn_context (TurnContext): Provides context for the current turn of conversation.
        """
        # Get the user's information (e.g., name, ID) from the incoming activity.
        user = turn_context.activity.from_property

        # Construct a mention text using the user's name, encoded for HTML.
        mention_text = f"<at>{escape(user.name)}</at>"
        # Create a mention entity with the user's details and the mention text.
        mention_entity = Mention(mentioned=user, text=mention_text, type="mention")

        # Create a text message that includes the mention text.
        reply_activity = MessageFactory.text(f"Hi {mention_text}")
        # Attach the mention entity to the message.
        reply_activity.entities = [mention_entity]

        # Send the message with the mention to the user.
        await turn_context.send_activity(reply_activity)

